File: leetcode/number_of_islands/number_of_islands.java
class Solution {

	public HashSet<List<Integer>> getEarthCoordSet(char[][] grid){
		HashSet<List<Integer>> earthCoordSet = new HashSet();

		for (Integer line = 0; line < grid.length; line++) {
			for (Integer col = 0; col < grid[0].length; col++) {
				if (grid[line][col] == '1') {
					List<Integer> earthCoord = List.of(line, col);
					earthCoordSet.add(earthCoord);
				}
			}
		}
		return earthCoordSet;
	}
	public int numIslands(char[][] grid) {
		Integer last_line = grid.length - 1;
		Integer last_col = grid[0].length - 1;
		HashSet<List<Integer>> earthCoordSet = this.getEarthCoordSet(grid);

		int islands = 0;
		while (! earthCoordSet.isEmpty()) {

			++islands;

			List<Integer> newIslandCoord = List.of();
			//System.out.println(earthCoordSet);
			for (var coord : earthCoordSet) {
				//System.out.println(coord);
				newIslandCoord = coord;
				earthCoordSet.remove(coord);
				break;
			}

			if (earthCoordSet.isEmpty())
				return islands;

			LinkedList<List<Integer>> newIslandCoords = new LinkedList();
			newIslandCoords.add(newIslandCoord);

			while (! newIslandCoords.isEmpty()) {
				
				List<Integer> coord = newIslandCoords.poll();
				Integer line = coord.get(0);
				Integer col = coord.get(1);

				List<Integer> upperCoord = List.of(line + 1, col);
				if((line + 1 <= last_line) && (earthCoordSet.contains(upperCoord))) {
					System.out.println(coord);
					newIslandCoords.add(upperCoord);
					earthCoordSet.remove(upperCoord);
				}

				List<Integer> bottomCoord = List.of(line - 1, col);
				if((line - 1 >= 0) && (earthCoordSet.contains(bottomCoord))) {
					System.out.println(coord);
					newIslandCoords.add(bottomCoord);
					earthCoordSet.remove(bottomCoord);
				}

				List<Integer> leftCoord = List.of(line, col - 1);
				if((col - 1 >= 0) && (earthCoordSet.contains(leftCoord))) {
					System.out.println(coord);
					newIslandCoords.add(leftCoord);
					earthCoordSet.remove(leftCoord);
				}

				List<Integer> rightCoord = List.of(line, col + 1);
				if((col + 1 <= last_col) && (earthCoordSet.contains(rightCoord))) {
					System.out.println(coord);
					newIslandCoords.add(rightCoord);
					earthCoordSet.remove(rightCoord);
				}
			}
		}
		return islands;
	}
}

File: leetcode/number_of_islands/number_of_islands.py
class Solution:
    def list_to_set(self, grid):
        set_grid = set({})
        for i in range(len(grid)):
            for u in range(len(grid[i])):
                if (grid[i][u] == "1"):
                    set_grid.add((i,u))

        return set_grid

    def numIslands(self, grid: List[List[str]]) -> int:
        islands = 0
        last_line = len(grid) - 1 
        last_col = len(grid[0]) - 1
        grid = self.list_to_set(grid)

        while(not (not grid)):
            new_island_point = grid.pop()
            islands += 1

            if (not grid):
                return islands

            new_island_points = [new_island_point]
            while (not (not new_island_points)):
                line, col = new_island_points.pop()

                if ((line + 1 <= last_line) and ((line + 1, col) in grid)):
                    new_island_points += [(line+1, col)]
                    grid.remove((line+1,col))

                if ((line - 1 >= 0) and ((line - 1, col) in grid)):
                    new_island_points += [(line-1, col)]
                    grid.remove((line-1,col))

                if ((col - 1 >= 0) and ((line, col - 1) in grid)):
                    new_island_points += [(line, col-1)]
                    grid.remove((line,col-1))

                if ((col + 1 <= last_col) and ((line, col + 1) in grid)):
                    new_island_points += [(line, col + 1)]
                    grid.remove((line,col+1))

        return islands
